include/sparrow/buffer_adaptor.hpp

```diff
@@ -0,0 +1,526 @@
+        constexpr const_reference operator[](size_type idx) const;
+
+        constexpr reference front();
+        constexpr const_reference front() const;
+
+        constexpr reference back();
+        constexpr const_reference back() const;
+
+        // Iterators
+
+        constexpr iterator begin() noexcept;
+        constexpr iterator end() noexcept;
+
+        constexpr const_iterator begin() const noexcept;
+        constexpr const_iterator end() const noexcept;
+
+        constexpr const_iterator cbegin() const noexcept;
+        constexpr const_iterator cend() const noexcept;
+
+        constexpr reverse_iterator rbegin() noexcept;
+        constexpr reverse_iterator rend() noexcept;
+
+        constexpr const_reverse_iterator rbegin() const noexcept;
+        constexpr const_reverse_iterator rend() const noexcept;
+
+        constexpr const_reverse_iterator crbegin() const noexcept;
+        constexpr const_reverse_iterator crend() const noexcept;
+
+        // Capacity
+
+        [[nodiscard]] constexpr size_type size() const noexcept;
+        [[nodiscard]] constexpr size_type max_size() const noexcept;
+        [[nodiscard]] constexpr size_type capacity() const noexcept;
+        [[nodiscard]] constexpr bool empty() const noexcept;
+        constexpr void reserve(size_type new_cap);
+        constexpr void shrink_to_fit();
+
+        // Modifiers
+
+        constexpr void clear() noexcept;
+
+        constexpr iterator insert(const_iterator pos, const value_type& value);
+        constexpr iterator insert(const_iterator pos, size_type count, const value_type& value);
+        template <class InputIt>
+            requires std::input_iterator<InputIt>
+        constexpr iterator insert(const_iterator pos, InputIt first, InputIt last);
+        constexpr iterator insert(const_iterator pos, std::initializer_list<value_type> ilist);
+
+        template <class... Args>
+        constexpr iterator emplace(const_iterator pos, Args&&... args);
+
+        constexpr iterator erase(const_iterator pos);
+        constexpr iterator erase(const_iterator first, const_iterator last);
+
+        constexpr void push_back(const value_type& value);
+
+        constexpr void pop_back();
+
+        constexpr void resize(size_type new_size);
+        constexpr void resize(size_type new_size, const value_type& value);
+
+    private:
+
+        static constexpr double m_from_to_size_ratio = static_cast<double>(sizeof(buffer_reference_value_type))
+                                                       / static_cast<double>(sizeof(value_type));
+
+        static constexpr size_type m_to_from_size_ratio = static_cast<size_type>(1. / m_from_to_size_ratio);
+
+        [[nodiscard]] constexpr buffer_reference::size_type index_for_buffer(size_type idx) const noexcept;
+
+        [[nodiscard]] buffer_reference::const_iterator get_buffer_reference_iterator(const_iterator pos);
+
+        buffer_reference& m_buffer;
+        const size_type m_max_size;
```

Comment:
The `const` makes this type non-copyable/movable. Do we really want that?
If yes, then add copy/move operations explicitly deleted so that it's clear to the user.

Reply:
Note also the tradefoff here of using memory instead of quick calculation (from a function), I'm not sure what's the best for our users. But it's easy to change later if needed (we are not ABI-sensible).

Reply:
I made the member mutable 👍